### ISSUE TITLE
vscode-neovim becomes unresponsive but nvim did *not* crash/disconnect

### ISSUE THREAD

**Issue by ian-h-chamberlain** [vscode-neovim/vscode-neovim#2136]
To be clear: this is a separate issue from #2000 and similar, because no "neovim disconnected" errors is displayed but the extension otherwise stops working. There are many similar / duplicate issues about neovim 0.10.0 crashing, please see those for workarounds etc.

---

Ok, last night I was reproducing this more consistently, on both `0.9.5` and `0.10.0`. Here are some of my notes about it, maybe they will help:

* No error message / notification is displayed when the problem occurs. Most of the time I was simply stuck in normal mode and unable to move the cursor or change modes, etc. 
  * In one case, I was stuck in macro-recording mode.
  * I have seen insert-mode before, but not typically
* Keybinds become unresponsive, but the extension is still active (logging, etc.) and the `neovim` process is still running:
   ```
   0.0  0.4  0:51.00 │  ├─ /Users/ianchamberlain/Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Plugin).app/Contents/MacOS/Code Helper (Plugin) --type=utility --utility-sub-type=node.mojom.NodeService --lang=en-US --service-sandbox-type=none --dns-resu
   0.0  0.1  0:00.00 │  │  └─ /opt/homebrew/bin/nvim -N --embed --cmd source /Users/ianchamberlain/.vscode/extensions/asvetliakov.vscode-neovim-1.17.2/runtime/vscode-neovim.vim
   ```
* `Developer: Toggle Keyboard Shortcuts Troubleshooting` showed it was still running `vscode-neovim.send` when pressing e.g. `hjkl`, so the context keys must have been set as for normal mode.

Extension logs (I believe the problem started around `2024-06-30T03:15:23.949Z`): [vscode-app-1719717602613.log](https://github.com/user-attachments/files/16046393/vscode-app-1719717602613.log)

Reproduction steps that were working for me almost 100% of the time, but don't seem to be working anymore. I'm not sure what changed, but this is basically what I had at the time:
1. 10-15 tabs open total, including several `:help` tabs (external buffers) 
2. search for phrase in workspace using VSCode shortcut (<kbd>cmd+shift+f</kbd> on macOS)    3. click a search result that opens a new editor (buffer that was not previously open)
4. click in the new editor
5. Observe stuck cursor / unresponsive, run `Neovim: Restart Extension` to go back to step 1

I don't have too many other clues yet, but maybe there's something there that helps... I'll keep trying to figure it out!

_Originally posted by @ian-h-chamberlain in https://github.com/vscode-neovim/vscode-neovim/issues/2131#issuecomment-2198786598_

**Comment by t-aida-yamato** [comment]
I'm new, but is it related to https://luals.github.io/ which is down?

**Comment by stefan-bc-ts** [comment]
I am getting these as well, today I was spamming closing tabs: Cmd W and got the neovim crashed message

**Comment by ian-h-chamberlain** [comment]
@stefan-bc-ts just to clarify: if you see a message about Neovim crashing, you are _not_ experiencing this issue, but most likely another instance of #2131 / #2000 (upstream neovim crash). See those issues for more details / possible fixes.

This issue is specifically about `vscode-neovim` becoming unresponsive _without_ crashing — the popup about the disconnect/crash never happens but movement controls stop working.

**Comment by ollien** [comment]
I'd be really curious to know if this occurs to anyone on a platform other than macOS. Anecdotally, I haven't experienced this on my personal Linux machine, but have experienced it on the Mac I use for work.

**Comment by ian-h-chamberlain** [comment]
Ok, now with the new progress messages I reproduced again! It seems to be stuck on this step:

![Screenshot 2024-07-08 at 10 24 24](https://github.com/vscode-neovim/vscode-neovim/assets/11131775/bf28ea03-57c6-40c3-a4ea-1206eeef9745)

Looking at #2137 I'll see if I can add some more logs/progress messages around those areas and get more details, but at least there's a starting point for where to look now.

Re: reproducing on different devices, I'm not sure if it's OS-dependent, but also anecdotally I tend to experience this more on my older Intel Mac for work than my personal M3 - but if it's timing-related, that could just be due to stuff being slower or something, I suppose.

**Comment by vlwkaos** [comment]
Am I experiencing this as well?, mine stuck on normal mode (cursor shows), but works as insert (vscode own editor). 
At this point neovim is practically dead and I need to restart it.

**Comment by ian-h-chamberlain** [comment]
@vlwkaos possibly! Do you see the same progress message I posted in my previous comment? I think I have occasionally experienced this getting stuck in "insert mode" although I'm not sure if the cursor was still visible in that case. I suppose if it happened while the mode was changing or something it could result in the kind of situation you're describing, and the symptom sounds the same.

Have you noticed any other situations that seem to make it worse / better? Even with the steps I listed above it's not super consistent for me, so it would be great if we could find a 100% reproduction scenario

**Comment by justinmk** [comment]
Just a guess, maybe https://github.com/vscode-neovim/vscode-neovim/issues/2242 is related? Could you try using the node-client from https://github.com/neovim/node-client/pull/395 and see if that helps? Or if that's not convenient, we can just update node-client and then try :)

**Comment by ian-h-chamberlain** [comment]
> Just a guess, maybe #2242 is related? Could you try using the node-client from [neovim/node-client#395](https://github.com/neovim/node-client/pull/395) and see if that helps? Or if that's not convenient, we can just update node-client and then try :)

Interesting, I don't think I ever saw any exception log like that when I was debugging this issue, but maybe the nature of the problem means they wouldn't normally be shown in the `vscode-neovim logs` output pane anyway? In any case, having both #2230 and this change should make the extension more resilient to the problem — I'll try to report back here after a while if I still see the requests hanging for a long time (which should be obvious by the new log I added).

**Comment by bierundboeller** [comment]
I get the unresponsiveness of neovim window in vscode error quite often as well - but only after using the VSCode Github Copilot extension, e.g. the chat window or the inline copilot pop-up. Restarting the vscode-neovim extension fixes the issue typically until the next usage of Copilot only. When I restart VSCode it usually takes some time until the error occurs again.

**Comment by justinmk** [comment]
> only after using the VSCode Github Copilot extension, e.g. the chat window or the inline copilot pop-up

That is tracked in https://github.com/vscode-neovim/vscode-neovim/issues/2184

The latest release v1.18.13 should help.

**Comment by mez0ru** [comment]
I still have this error with the latest version, I didn't even use vscode that long until I had it. It shows the notification"syncing editors". I don't remember the exact notification message was, but it never worked after that. I had to close vscode.

**Comment by xiyaowong** [comment]
Seems like no one has reported this issue recently, so I'll close it for now.

**Comment by xiyaowong** [comment]
@BenMcH  Please confirm if your issue is similar to #2407. I haven't encountered this issue recently.

**Comment by BenMcH** [comment]
Oh yeah, I didn't realize they were separate issues. #2407 is the issue I'm encountering

**Comment by ian-h-chamberlain** [comment]
As far as I know this issue _is_ different from #2407 since I originally reproduced it on 0.9.5, but I haven't noticed it in a while due to that issue happening a lot more frequently.

If/when #2407 is fixed perhaps this issue will start being more obviously reproducible, or if not then it's probably safe to close after some time.

**Comment by vivsamInstr** [comment]
This is happening for me when I navigate to another file by ctrl+clicking on the variable name (go to definition). The neo vim controls (including jklh) stops working. The VScode throws an error message as in the attached image and I have to restart the code to make the neovim work. Hope this helps in troubleshooting the issues or let me know if I need to log this as a new issue

<img width="458" alt="Image" src="https://github.com/user-attachments/assets/9a73f5b2-24e6-450c-8554-267e2f350263" />

**Comment by vivsamInstr** [comment]
Full error dump

<img width="368" alt="Image" src="https://github.com/user-attachments/assets/60583580-ecdf-4ff8-833b-3a373df2572f" />

**Comment by xiyaowong** [comment]
@vivsamInstr Not the same bug. You can create a new issue.

**Comment by biel-correa** [comment]
<img width="459" alt="Image" src="https://github.com/user-attachments/assets/92d40f54-faca-49d0-a1ac-3fbbef4a8893" />

Still having the same issue. The logs didn't show anything usefull

I'm on MacOs

**Comment by ollien** [comment]
> <img alt="Image" width="459" src="https://private-user-images.githubusercontent.com/56176344/430915824-92d40f54-faca-49d0-a1ac-3fbbef4a8893.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQwMjcyNzYsIm5iZiI6MTc0NDAyNjk3NiwicGF0aCI6Ii81NjE3NjM0NC80MzA5MTU4MjQtOTJkNDBmNTQtZmFjYS00OWQwLWExYWMtM2ZiYmVmNGE4ODkzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDA3VDExNTYxNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI5Yjc3YzhlMjZkYzhlMjFiODdjYmE2MGI2MWQ4OGMwMGQyODVlM2RjZmQwNDBlOTMwY2I4ZDZlZGRmMTc3YTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.OxzQq-iwXDaBBsM2zF7He_frwKf0hgGe-Iy-FdZ4G98">
> 
> Still having the same issue. The logs didn't show anything usefull
> 
> I'm on MacOs

This is #2407

**Comment by shelper** [comment]
same issue, and this is what i got in vscode neovim log
2025-04-09 23:53:30.868 [warning] BufferManager: Cleanup took longer than 5s:  { windows: [ 1015 ], buffers: [ 11 ] }

**Comment by chrisjaquet** [comment]
> I'd be really curious to know if this occurs to anyone on a platform other than macOS. Anecdotally, I haven't experienced this on my personal Linux machine, but have experienced it on the Mac I use for work.

I experience this on Windows, also usually when I have multiple tabs (~10) open.

VSCode 1.99.2 (user setup)
asvetliakov.vscode-neovim 1.18.19
NVIM v0.10.1
Windows_NT x64 10.0.19045

**Comment by justinmk** [comment]
Fixed by https://github.com/neovim/neovim/pull/34412 in Nvim 0.12 (unreleased), and will be included in Nvim 0.11.3 maintenance release.
